CMakeLists.txt

```diff
@@ -778,45 +778,82 @@ function(add_library_variants)
+    armv4t          ""                  "-march=armv4t"                                         "target=armv4t-none-unknown-eabi mfpu=none"                                                                "-M musicpal -cpu arm926"
+    armv5te         ""                  "-march=armv5te"                                        "target=armv5e-none-unknown-eabi mfpu=none"                                                                "-M musicpal -cpu arm926"
+    armv6m          soft_nofp           "-mfloat-abi=soft -march=armv6m"                        "target=thumbv6m-none-unknown-eabi mfpu=none"                                                              "-M microbit"
+    armv7m          soft_nofp           "-mfloat-abi=soft -march=armv7m+nofp"                   "target=thumbv7m-none-unknown-eabi mfpu=none"                                                              "-M mps2-an385 -cpu cortex-m3"
+    armv7em         soft_nofp           "-mfloat-abi=soft -march=armv7em -mfpu=none"            "target=thumbv7em-none-unknown-eabi mfpu=none"                                                             "-M mps2-an386 -cpu cortex-m4"
+    armv7em         hard_fpv4_sp_d16    "-mfloat-abi=hard -march=armv7em -mfpu=fpv4-sp-d16"     "target=thumbv7em-none-unknown-eabihf mfpu=fpv4-sp-d16 mfloat-abi=hard"                                    "-M mps2-an386 -cpu cortex-m4"
+    armv7em         hard_fpv5_d16       "-mfloat-abi=hard -march=armv7em -mfpu=fpv5-d16"        "target=thumbv7em-none-unknown-eabihf mfpu=fpv5-d16 mfloat-abi=hard"                                       "-M mps2-an500 -cpu cortex-m7"
+    armv8m.main     soft_nofp           "-mfloat-abi=soft -march=armv8m.main+nofp"              "target=thumbv8m.main-none-unknown-eabi mfpu=none"                                                         "-M mps2-an505 -cpu cortex-m33"
+    armv8m.main     hard_fp             "-mfloat-abi=hard -march=armv8m.main+fp"                "target=thumbv8m.main-none-unknown-eabihf mfpu=fpv5-d16 mfloat-abi=hard"                                   "-M mps3-an524 -cpu cortex-m33"
+    armv8.1m.main   soft_nofp_nomve     "-mfloat-abi=soft -march=armv8.1m.main+nofp+nomve"      "target=thumbv8.1m.main-none-unknown-eabi mfpu=none"                                                       "-M mps3-an547 -cpu cortex-m55"
+    armv8.1m.main   hard_fp             "-mfloat-abi=hard -march=armv8.1m.main+fp"              "target=thumbv8.1m.main-none-unknown-eabihf march=+fp16 mfpu=fp-armv8-fullfp16-sp-d16 mfloat-abi=hard"     "-M mps3-an547 -cpu cortex-m55"
+    armv8.1m.main   hard_nofp_mve       "-mfloat-abi=hard -march=armv8.1m.main+nofp+mve"        "target=thumbv8.1m.main-none-unknown-eabihf march=+dsp march=+mve mfpu=none mfloat-abi=hard"               "-M mps3-an547 -cpu cortex-m55"
+)
+
+configure_file(
+    ${CMAKE_CURRENT_SOURCE_DIR}/cmake/multilib.yaml.in
+    ${CMAKE_CURRENT_BINARY_DIR}/multilib.yaml
```

Comment:
With multilib.yaml in this directory I think it will be inactive if you run clang from the build directory. This would result in a confusing difference of behaviour between the locally built clang and the installed clang. I suggest putting multilib.yaml in `lib/clang-runtimes` next to the libraries.

IIRC this will break tests but I think cherry-picking the following changes will address that:
* https://reviews.llvm.org/D143590
* https://reviews.llvm.org/D143666
* https://reviews.llvm.org/D144542

Reply:
I had an entertaining problem with this suggestion.

If you drop `multilib.yaml` into a directory where clang can find it when run from the build tree, then all the picolibc builds break, because the picolibc built-in multilib support activates!

`meson.build` tries to run `clang` with `--print-multi-lib`, and if it gets a list of variants back, it automatically makes a build directory for each one. But _this_ CMakeLists.txt is also setting up a different build directory for each one, so you end up trying to build N libraries in each of those N subdirs, for N^2 libraries in total, N(N-1) of which had two sets of contradictory options smushed on to their command lines and get very confused.

Perhaps in future we should consider letting picolibc do its thing all by itself, but for the moment, I'll do the smallest change that gets things working again: explicitly turn picolibc's multilib feature off on the meson command line.